In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn
import warnings; warnings.simplefilter('ignore')

In [ ]:

book_data = pd.read_csv("/content/Books.csv")
book_data.head()


ISBN  \
0  0195153448   
1  0002005018   
2  0060973129   
3  0374157065   
4  0393045218   

                                                                                           Book-Title  \
0  Classical Mythology                                                                                  
1  Clara Callan                                                                                         
2  Decision in Normandy                                                                                 
3  Flu: The Story of the Great Influenza Pandemic of 1918 and the Search for the Virus That Caused It   
4  The Mummies of Urumchi                                                                               

            Book-Author Year-Of-Publication                   Publisher  \
0  Mark P. O. Morford    2002                Oxford University Press      
1  Richard Bruce Wright  2001                HarperFlamingo Canada        
2  Carlo D'Este          1991                HarperPerennial              
3  Gina Bari Kolata      1999                Farrar Straus Giroux         
4  E. J. W. Barber       1999                W. W. Norton &amp; Company   

                                                    Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.01.THUMBZZZ.jpg   
1  http://images.amazon.com/images/P/0002005018.01.THUMBZZZ.jpg   
2  http://images.amazon.com/images/P/0060973129.01.THUMBZZZ.jpg   
3  http://images.amazon.com/images/P/0374157065.01.THUMBZZZ.jpg   
4  http://images.amazon.com/images/P/0393045218.01.THUMBZZZ.jpg   

                                                    Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.01.MZZZZZZZ.jpg   
1  http://images.amazon.com/images/P/0002005018.01.MZZZZZZZ.jpg   
2  http://images.amazon.com/images/P/0060973129.01.MZZZZZZZ.jpg   
3  http://images.amazon.com/images/P/0374157065.01.MZZZZZZZ.jpg   
4  http://images.amazon.com/images/P/0393045218.01.MZZZZZZZ.jpg   

                                                    Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.01.LZZZZZZZ.jpg  
1  http://images.amazon.com/images/P/0002005018.01.LZZZZZZZ.jpg  
2  http://images.amazon.com/images/P/0060973129.01.LZZZZZZZ.jpg  
3  http://images.amazon.com/images/P/0374157065.01.LZZZZZZZ.jpg  
4  http://images.amazon.com/images/P/0393045218.01.LZZZZZZZ.jpg

In [ ]:
book_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 8 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271360 non-null  object
 1   Book-Title           271360 non-null  object
 2   Book-Author          271359 non-null  object
 3   Year-Of-Publication  271360 non-null  object
 4   Publisher            271358 non-null  object
 5   Image-URL-S          271360 non-null  object
 6   Image-URL-M          271360 non-null  object
 7   Image-URL-L          271357 non-null  object
dtypes: object(8)
memory usage: 16.6+ MB


In [ ]:
book_data.shape

(271360, 8)

In [ ]:
users_data= pd.read_csv('/content/Users.csv')
users_data.head()

User-ID                            Location   Age
0  1        nyc, new york, usa                 NaN  
1  2        stockton, california, usa           18.0
2  3        moscow, yukon territory, russia    NaN  
3  4        porto, v.n.gaia, portugal           17.0
4  5        farnborough, hants, united kingdom NaN

In [ ]:
users_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   User-ID   278858 non-null  int64  
 1   Location  278858 non-null  object 
 2   Age       168096 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 6.4+ MB


In [ ]:
users_data.shape

(278858, 3)

**Ratings Data**

In [ ]:
ratings_data = pd.read_csv("/content/Ratings.csv")
ratings_data.head()

User-ID        ISBN  Book-Rating
0  276725   034545104X  0          
1  276726   0155061224  5          
2  276727   0446520802  0          
3  276729   052165615X  3          
4  276729   0521795028  6

In [ ]:
ratings_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1149780 entries, 0 to 1149779
Data columns (total 3 columns):
 #   Column       Non-Null Count    Dtype 
---  ------       --------------    ----- 
 0   User-ID      1149780 non-null  int64 
 1   ISBN         1149780 non-null  object
 2   Book-Rating  1149780 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 26.3+ MB


In [ ]:
ratings_data.shape

(1149780, 3)

In [ ]:

n_users = train_data['u_unique'].nunique()
n_books = train_data['i_unique'].nunique()

train_matrix = np.zeros((n_users, n_books))

for entry in train_data.itertuples():
    train_matrix[entry[1]-1, entry[2]-1] = entry[3]

In [ ]:
train_matrix.shape

(54182, 81054)

In [ ]:
n_users = test_data['u_unique'].nunique()
n_books = test_data['i_unique'].nunique()

test_matrix = np.zeros((n_users, n_books))

for entry in test_data.itertuples():
    test_matrix[entry[1]-1, entry[2]-1] = entry[3]

In [ ]:
test_matrix.shape

(23117, 33687)

In [ ]:

train_matrix_small = train_matrix[:5000, :5000]
test_matrix_small = test_matrix[:5000, :5000]

from sklearn.metrics.pairwise import pairwise_distances
user_similarity = pairwise_distances(train_matrix_small, metric='cosine')
item_similarity = pairwise_distances(train_matrix_small.T, metric='cosine')

In [ ]:

def predict_books(ratings, similarity, type='user'):
    if type == 'user':
        mean_user_rating = ratings.mean(axis=1)


        ratings_diff = (ratings - mean_user_rating[:, np.newaxis])
        pred = mean_user_rating[:, np.newaxis] + similarity.dot(ratings_diff) / np.array([np.abs(similarity).sum(axis=1)]).T
    elif type == 'item':
        pred = ratings.dot(similarity) / np.array([np.abs(similarity).sum(axis=1)])
    return pred

In [ ]:
item_prediction = predict_books(train_matrix_small, item_similarity, type='item')
user_prediction = predict_books(train_matrix_small, user_similarity, type='user')

In [ ]:

from sklearn.metrics import mean_squared_error
from math import sqrt

def rmse(prediction, test_matrix):
    prediction = prediction[test_matrix.nonzero()].flatten()
    test_matrix = test_matrix[test_matrix.nonzero()].flatten()
    return sqrt(mean_squared_error(prediction, test_matrix))

print(f'Item-based CF RMSE: {rmse(item_prediction, test_matrix_small)}')
print(f'User-based CF RMSE: {rmse(user_prediction, test_matrix_small)}')

Item-based CF RMSE: 7.962532654166158
User-based CF RMSE: 7.962186775709137


In [ ]:
from surprise import Reader, Dataset

In [ ]:

reader = Reader(rating_scale=(1, 10))
data = Dataset.load_from_df(ratings_explicit, reader)

In [ ]:
from surprise import SVD, model_selection, accuracy
model = SVD()


%time model_selection.cross_validate(model, data, measures=['RMSE'], cv=5, verbose=True)

Evaluating RMSE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.6332  1.6350  1.6292  1.6302  1.6302  1.6316  0.0022  
Fit time          7.26    6.78    6.35    7.12    7.03    6.91    0.32    
Test time         1.01    1.21    0.50    0.95    1.09    0.95    0.24    
CPU times: user 43.8 s, sys: 425 ms, total: 44.2 s
Wall time: 44.5 s


{'test_rmse': array([1.63319233, 1.63502867, 1.62917541, 1.63022798, 1.63016579]),
 'fit_time': (7.255144119262695,
  6.7794349193573,
  6.34954309463501,
  7.120924472808838,
  7.026201248168945),
 'test_time': (1.0149495601654053,
  1.206634759902954,
  0.4985623359680176,
  0.947761058807373,
  1.0928750038146973)}

## Train - Test Split

In [ ]:

trainset, testset = model_selection.train_test_split(data, test_size=0.2)


model = SVD()
model.fit(trainset)

## Evaluation metrics for SVD model

In [ ]:

predictions = model.test(testset)
print(f"The accuracy is {accuracy.rmse(predictions)}")

RMSE: 1.6276
The accuracy is 1.6275624268787283


The RMSE score is better now with 1.63.

## Testing Results

In [ ]:

uid = 276744
iid = '038550120X'
pred = model.predict(uid, iid, verbose=True)

user: 276744     item: 038550120X r_ui = None   est = 7.45   {'was_impossible': False}


Now let's display estimated rating and real rating

In [ ]:

print(f'The estimated rating for the book with ISBN code {pred.iid} from user #{pred.uid} is {pred.est:.2f}.\n')
actual_rtg= ratings_explicit[(ratings_explicit.user_id==pred.uid) &
                             (ratings_explicit.isbn==pred.iid)].book_rating.values[0]
print(f'The real rating given for this was {actual_rtg:.2f}.')

The estimated rating for the book with ISBN code 038550120X from user #276744 is 7.45.

The real rating given for this was 7.00.


In [ ]:

from collections import defaultdict

def get_top_n(predictions, n=10):

    top_n = defaultdict(list)
    for uid, iid, true_r, est, _ in predictions:
        top_n[uid].append((iid, est))

    for uid, user_ratings in top_n.items():
        user_ratings.sort(key=lambda x: x[1], reverse=True)
        top_n[uid] = user_ratings[:n]

    return top_n

In [ ]:

pred = model.test(testset)
top_n = get_top_n(pred)

In [ ]:
pred_testset_df = pd.DataFrame(pred)
pred_testset_df

uid         iid  r_ui       est                    details
0      16106   0449213358  10.0  8.407405  {'was_impossible': False}
1      276231  0425163407  7.0   6.237226  {'was_impossible': False}
2      191913  0446612545  9.0   7.614947  {'was_impossible': False}
3      193529  0345356748  8.0   7.244047  {'was_impossible': False}
4      118320  0451156609  2.0   6.756214  {'was_impossible': False}
...       ...         ...  ...        ...                        ...
63219  258151  0671027581  8.0   7.703824  {'was_impossible': False}
63220  237035  0140234217  9.0   7.478401  {'was_impossible': False}
63221  202127  0916441237  8.0   7.855660  {'was_impossible': False}
63222  242528  0679731148  7.0   8.034314  {'was_impossible': False}
63223  256470  0451184718  8.0   8.098775  {'was_impossible': False}

[63224 rows x 5 columns]

In [ ]:
def get_reading_list(userid):

    reading_list = defaultdict(list)
    top_n = get_top_n(pred, n=10)
    for n in top_n[userid]:
        book, rating = n
        title = new_book_df.loc[new_book_df.isbn==book].book_title.unique()[0]
        reading_list[title] = rating
    return reading_list

In [ ]:

example_reading_list = get_reading_list(userid=116866)
for book, rating in example_reading_list.items():
    print(f'{book}: {rating}')

Climbing Mount Improbable: 7.698358001645088
Metamagical Themas: Questing for the Essence of Mind and Pattern: 7.660396502374004
